Remove hashtags from the last line of a status if it only contains hashtags

renchap opened the pull request:
This should have the same effect as #26260, but done in the frontend directly.

It avoids removing the hashtags from the original status text, as this breaks compatibility with older clients and instance which do not (yet) support displaying out-of-band hashtags.

Once enough of the ecosystem moved to display out-of-band hashtags properly, then we can probably switch to removing them from the content at the time of posting.

I tried to comment the code so it is understandable, and it uses proper Typescript (except for the state types that I needed to work around, as our state is not-yet typed). There are also tests covering the cases I could think of.

I chose to have a helper function do the work, so we only need to parse the content once, and the results are used in 2 places (the hashtag bar and `<StatusContent>`). It allows the code handling the hashtags to be co-located.

Cases that I handled:
- full out-of-band hashtags (not included in the status content)
- if the last line of a status is only hashtags, they are moved to the dedicated bar
- if a hashtag in the last line appears in the content, then is it not added to the bar
- case variations do not matter (hashtags are all lowercase internally, but may have different cases in the post)
- but if there are case variations of the same hashtag (for example in the last line), then we choose one of those only to be in the hashtag bar. I chose to go with the one having the most difference between upper and lower case letters, to try to select the most camel-cased one (this is important for accessibility)
- if there is only one line with only hashtags in the content, and the status has a media, then we do not show any content and  display the hashtags in the bar. But not if the status has a CW, otherwise the CW hides zero content and this is weird

Something that is not specifically handled is the order of the hashtags in the bar, but I am not sure of what to decide here.

Before:
<img width="592" alt="image" src="https://github.com/mastodon/mastodon/assets/42070/47312fbc-ec24-4928-8255-28107a6bccac">

After:
<img width="582" alt="image" src="https://github.com/mastodon/mastodon/assets/42070/d27efa1b-82e2-4af2-8255-b6d5c0d4cc80">


I wonder if this should not be done server-side, with a new API parameter to tell the server that you support displaying out-of-band hashtags, so the server can serve the already-trimmed text, but this is probably more complex to implement.

ClearlyClaire commented:
I'm still not 100% behind that idea but doing it client-side alleviates most of my issues with #26260. The approach looks fine except for the matching logic which is yet another variant where we probably want only one (see #26506)

renchap commented:
I will rework this PR to include your change, and avoid parsing the status multiple times

holmesrichards commented:
Is there some place where a discussion of this whole hashtags bar thing is going on?

The devs seem to be going all in on it, but I don't see the point of it, and it is not at all obvious to me that users welcome it. As I understand it the ultimate goal is to replace hashtags at the end of a post with a hashtag bar — which takes up more space, IMHO is visually more intrusive (with buttons that are larger and more visually emphasized than the hashtags), and yields no obvious functionality improvement.

renchap commented:
> hashtags that were not visible on posts at all because they were in the metadata

This has already been merged, this specific PR is about the second part (handling hashtags that are on the last line, if it only contains hashtags)

renchap committed:
Convert <HashtagBar> to Typescript

renchap committed:
Display hashtags in the dedicaced bar if they are in the last line, and this line only contains hashtags

renchap commented:
Pushed a new version which should handle more cases, and updated the PR description with new information

renchap committed:
Use `localeCompare` to match #26525

renchap commented on app/javascript/mastodon/components/hashtag_bar.tsx:
```diff
@@ -0,0 +1,195 @@
+ * @returns The input hashtags, but with only 1 occurence of each (case-insensitive)
+ */
+function uniqueHashtagsWithCaseHandling(hashtags: string[]) {
+  const groups = groupBy(hashtags, (tag) => tag.toLowerCase());
```
I am using `toLowerCase` here, as using `localeCompare` would be more involved (manually build a list for each new tag, and then compare every tag with every list's keys).

I am not sure this is really worth it, but I can do it if needed.

ClearlyClaire commented on app/javascript/mastodon/components/hashtag_bar.tsx:
```diff
@@ -0,0 +1,195 @@
+  // Now we parse the last line, and try to see if it only contains hashtags
+  const lastLineHashtags: string[] = [];
+  // try to see if the last line is only hashtags
+  const onlyHashtags = Array.from(lastChild.childNodes).every((node) => {
+    if (isNodeLinkHashtag(node)) {
+      const normalized = normalizeHashtag(node.innerText);
+
+      if (!localeAwareInclude(tagNames, normalized))
+        // stop here, this is not a real hashtag, so consider it as text
+        return false;
+
+      if (!localeAwareInclude(contentHashtags, normalized))
+        // only add it if it does not appear in the rest of the content
+        lastLineHashtags.push(normalized);
+
+      return true;
+    } else if (node.nodeType === Node.TEXT_NODE && !node.nodeValue?.trim()) {
+      // This is a space
+      return true;
+    } else return false;
+  });
```
Not all hashtags from the last child get added to `lastLineHashtags` if something ends up not being a hashtag: indeed, `every` will stop early.

This will cause the hashtag bar to include hashtags that are in the body of the post.

ClearlyClaire commented on app/javascript/mastodon/components/hashtag_bar.tsx:
```diff
@@ -0,0 +1,195 @@
+  );
+
+  if (onlyHashtags) {
+    statusContent = contentWithoutLastLine.innerHTML;
```
When there is only a line of tags, this is empty, and thus ignored, leading to the tags appear both in the bar and the contents.

renchap committed:
Add tests for the hashtag bar computation

renchap committed:
If a status has only hashtags on one line as content and a media, put the hashtags in the bar.

But do not do it if it has a CW, so the CW is not hiding zero content

renchap commented:
Updated with your remarks:
- single-line statuses now work
- accentuated hashtags with case differences are properly grouped
- statuses with media are handled properly, if the content is only a line of hashtags, then those are in the hashtag bar
- if the status has a CW, a media and one unique line of only hashtags, then those are not added to the hashtag bar to avoid having no content under the CW
- refactored the code a bit to add tests for all the use-cases I had in mind

ClearlyClaire reviewed (approved):
Still not completely sold on the idea, but if we're doing it, the code looks solid.

nclm commented:
Hi!

I see a lot of status where half of the ending hashtags are in-line and another half is in the new hashtag ribbon.

Two examples totally at random: 

![image](https://github.com/mastodon/mastodon/assets/923008/88ee1890-cfa9-4500-b856-0b7157f56b93)

https://mastodon.social/@joshisanonymous@h4.io/110271009190498547

![image](https://github.com/mastodon/mastodon/assets/923008/f6193311-f9af-492f-912f-50449c0ba26c)

https://mastodon.social/@dad@mastodon.eole.education/109795965672469290

Not sure why, but it looks like a bug of the new system to me?

renchap commented:
The version containing this change has not yet been deployed to mastodon.social. We will most likely deploy it tomorrow.

nclm commented:
Ah sorry, I thought this was where the new hashtag ribbon was introduced :) Looking forward to see that new version!

nclm commented:
Update: this still happens with v4.2.0+nightly-2023-08-22.

renchap commented:
Thanks, looking at the JSON for this status shows that there are `<span>` inside the hashtags links, and this is not something that I handled. I will try to reproduce and fix it.

There may be an issue with accentuated characters as well. The server normalizes the `tag` attributes, and I may not replicate this correctly in my code.

You can open an issue to track this if you want.

renchap commented:
#26614 should fix it

trankten commented:
Hello.

I want to point out that for some reason these hashtags are not displaying sepparately. The message was posted using the API statuses POST in case it helps.

Check: https://tkz.one/@knfansub/111107902034494236

renchap commented:
They are not in their own paragraph

trankten commented:
> They are not in their own paragraph

Oh, thanks for the clarification. 👍  My bad.
